mcp sse stateful session in contrib

jue-yin opened the pull request:
Additional Description:
Sync the feature in envoy-proxy.

This PR adds native support for the [MCP 241105 spec](https://modelcontextprotocol.io/specification/2024-11-05/basic/transports#http-with-sse) in the envelope stateful session extension.
It enables Envoy to handle session context in SSE event streams, allowing session ID and upstream host to be encoded/decoded as required by the protocol.

Separator Description:
In SSE mode, the session ID and upstream host are joined using a separator . (dot).

The format is: sessionId=<original_session_id>.<base64url_encoded_host>
For example: sessionId=abcdefg.MTI3LjAuMC4xOjgw
When processing requests, Envoy splits the value at the last . to extract the original session ID and the upstream host (after Base64Url decoding).
The implementation covers both header-based and SSE (query param) session modes, and refactors the interface for better extensibility and clarity.
Comprehensive unit and integration tests are included to ensure correctness, including strict mode and various edge cases.

Risk Level:
Low

Testing:
Unit and integration tests added and enhanced for new SSE session mode.

Docs Changes:
Updated relevant documentation to describe the new SSE session mode, separator usage, and example formats.

Release Notes:
Added

Platform Specific Features:
N/A

Relevant Links:

Related issue: https://github.com/envoyproxy/envoy/issues/40247
Related previous PR: https://github.com/envoyproxy/envoy/pull/39004
[MCP 241105 Spec](https://modelcontextprotocol.io/specification/2024-11-05/basic/transports#http-with-sse)

recommand use route level Envoyfilter to 
istio EnvoyFilter example:

```yaml
  apiVersion: networking.istio.io/v1alpha3
  kind: EnvoyFilter
  metadata:
    name: mcp-sse-stateful-session
    labels:
      istio/cluster-id: {istio-pod-name}
  spec:
    configPatches:
      - applyTo: HTTP_FILTER
        match:
          context: GATEWAY
          listener:
            filterChain:
              filter:
                name: envoy.filters.network.http_connection_manager
                subFilter:
                  name: envoy.filters.http.router
        patch:
          operation: INSERT_BEFORE
          value:
            name: envoy.filters.http.mcp_sse_stateful_session
            typed_config:
              "@type": type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: type.googleapis.com/envoy.extensions.filters.http.mcp_sse_stateful_session.v3alpha.McpSseStatefulSession
  
      - applyTo: HTTP_ROUTE
        match:
          context: GATEWAY
          routeConfiguration:
            vhost: 
              route:
                name: {specific_route_name}
        patch:
          operation: MERGE
          value:
            typed_per_filter_config:
              envoy.filters.http.mcp_sse_stateful_session:
                "@type": type.gooleapis.com/udpa.type.v1.TypedStruct 
                type_url: type.googleapis.com/envoy.extensions.filters.http.mcp_sse_stateful_session.v3alpha.McpSseStatefulSessionPerRoute
                value:
                  mcp_sse_stateful_session:
                    session_state:
                      name: envoy.http.mcp_sse_stateful_session.envelope
                      typed_config:
                       "@type": type.googleapis.com/udpa.type.v1.TypedStruct
                       type_url: type.googleapis.com/envoy.extensions.http.mcp_sse_stateful_session.envelope.v3alpha.EnvelopeSessionState
                       value:
                         param_name: "sessionId"
                         chunk_end_patterns: ["\r\n\r\n", "\n\n", "\r\r"]
                    strict: true
```

jue-yin committed:
mcp sse stateful session in contrib

Signed-off-by: 爵银 <jueyin.hsl@alibaba-inc.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/higress-group/envoy?pullRequest=12) <br/>All committers have signed the CLA.

jue-yin committed:
fix generic_factory_context problem
cherry-pick strict mode from envoy proxy https://github.com/envoyproxy/envoy/pull/30573

Signed-off-by: jue-yin <jueyin.hsl@alibaba-inc.com>

jue-yin committed:
remove wrong file

Signed-off-by: jue-yin <jueyin.hsl@alibaba-inc.com>

jue-yin committed:
fix typo

Signed-off-by: 爵银 <jueyin.hsl@alibaba-inc.com>

johnlanni reviewed (approved):
LGTM
